File: backend/common/schemas/action.py
"""
Pydantic schemas for Action JSON envelopes.

Inherits standard bases. Add model-specific fields only.

Times & Billable (established 2026-09-09):
  action.times  — clock in/out array for any action, billable or not.
  action.billable — billing config, estimates, actuals, invoice linkage.
  Kept as separate top-level JSON fields on the Action model so that
  time tracking and billing are independent concerns.
"""
from __future__ import annotations

from typing import Optional
from pydantic import BaseModel, Field, field_validator, model_validator

from .envelopes import ConfigBase, MetadataBase, RecordPrefsBase, RefsBase, SourceRef


# ═══════════════════════════════════════════════════════════════════════
# action.times — clock in/out entries for any action
# ═══════════════════════════════════════════════════════════════════════

class TimeEntry(BaseModel):
    """One clock-in / clock-out span on an action."""
    id: Optional[str] = None                  # uuid — assigned on creation
    who: Optional[int] = None                 # contact_id of person logging time
    dt_in: Optional[int] = None               # epoch ms — clock in
    dt_out: Optional[int] = None              # epoch ms — clock out (null = still clocked in)
    elapsed_ms: Optional[int] = None          # dt_out - dt_in (computed on clock-out)
    percent_active: int = Field(
        100, ge=0, le=100,
        description="Utilization — 100 = fully productive, 50 = half idle/waiting",
    )
    reason: str = ""                          # what this time span was for
    notes: str = ""                           # freeform detail
    tags: list[str] = Field(default_factory=list)   # industry/task tags
    issue: Optional[str] = None               # interruption, delay, problem encountered

    class Config:
        extra = "allow"                       # industries will add fields we haven't thought of

    @model_validator(mode="after")
    def compute_elapsed(self) -> "TimeEntry":
        if self.dt_in is not None and self.dt_out is not None and self.elapsed_ms is None:
            self.elapsed_ms = self.dt_out - self.dt_in
        return self


class ActionTimes(BaseModel):
    """action.times — time tracking for any action regardless of billing."""
    entries: list[TimeEntry] = Field(default_factory=list)
    total_elapsed_ms: Optional[int] = None    # sum of entries[].elapsed_ms
    total_active_ms: Optional[int] = None     # sum adjusted by percent_active

    class Config:
        extra = "forbid"


# ═══════════════════════════════════════════════════════════════════════
# action.billable — billing configuration, estimates, actuals
# ═══════════════════════════════════════════════════════════════════════

class ActionBillable(BaseModel):
    """action.billable — the money side of an action."""
    # -- who is billed --
    customer_id: Optional[int] = None         # billing party (customer record)
    contact_id: Optional[int] = None          # specific person on the account
    company: str = ""                         # display name (denormalized)
    attention: str = ""                       # attention line on invoice

    # -- rate --
    is_billable: bool = True
    rate: Optional[float] = None              # per rate_unit
    rate_unit: str = Field(
        "hour",
        description="hour | day | flat | unit",
    )
    currency: str = "USD"

    # -- categorization --
    skill_category: str = ""                  # journeyman_plumber, electrician, etc.
    activity: str = ""                        # installation, service, consultation, etc.

    # -- estimates vs actuals --
    hours_estimated: Optional[float] = None
    hours_actual: Optional[float] = None
    total_estimated: Optional[float] = None   # rate * hours_estimated (or flat amount)
    total_actual: Optional[float] = None      # rate * hours_actual (or flat amount)
    variance_pct: Optional[float] = None      # ((actual - estimated) / estimated) * 100

    # -- product set (what is being installed / serviced) --
    product_set: list[int] = Field(
        default_factory=list,
        description="Item IDs that this labor applies to",
    )

    # -- invoice linkage --
    invoice_id: Optional[int] = None
    invoice_line: Optional[int] = None

    # -- Alice learning signal --
    estimate_source: Optional[str] = Field(
        None,
        description="Who produced the estimate: alice | user | historical",
    )
    estimate_confidence: Optional[int] = Field(
        None, ge=1, le=10,
        description="1-10 confidence in the estimate (Alice or user)",
    )

    class Config:
        extra = "allow"                       # industries will add fields


# -- .config ----------------------------------------------------------------

class ActionConfig(ConfigBase):
    times: Optional[ActionTimes] = None
    billable: Optional[ActionBillable] = None


# -- .metadata (inherits MetadataBase) --------------------------------------

class ActionMetadata(MetadataBase):
    pass


# -- .prefs (inherits RecordPrefsBase) --------------------------------------

class ActionPrefs(RecordPrefsBase):
    pass


# -- .refs (inherits RefsBase) ----------------------------------------------

class ActionRefs(RefsBase):
    tags: list[str] = Field(default_factory=list)
    keywords: list[str] = Field(default_factory=list)
    source: Optional[SourceRef] = None

File: frontend/src/apps/core/models/action/Action.ts
/* LastChecked: 2026-09-09 | WhereUsed: TODO(wc3-schema-audit) | WhoCreated: Claude */
// WC3 ↔ R25 TypeScript alignment interface for Action
// Mirrors Pydantic schemas: common/schemas/action.py

// ── action.config.times — clock in/out entries (mirrors ActionTimes) ───

/** One clock-in / clock-out span. Mirrors TimeEntry in action.py. */
export interface TimeEntry {
  id: string;                              // uuid
  who?: number | null;                     // contact_id of person logging time
  dt_in?: number | null;                   // epoch ms — clock in
  dt_out?: number | null;                  // epoch ms — clock out (null = still clocked in)
  elapsed_ms?: number | null;              // dt_out - dt_in (computed on clock-out)
  percent_active?: number;                 // 0-100, default 100
  reason?: string;                         // what this time span was for
  notes?: string;                          // freeform detail
  tags?: string[];                         // industry/task tags
  issue?: string | null;                   // interruption, delay, problem
  [key: string]: unknown;                  // extra="allow" — industry extensions
}

/** action.config.times — time tracking for any action. Mirrors ActionTimes. */
export interface ActionTimes {
  entries: TimeEntry[];
  total_elapsed_ms?: number | null;        // sum of entries[].elapsed_ms
  total_active_ms?: number | null;         // sum adjusted by percent_active
}

// ── action.config.billable — billing configuration (mirrors ActionBillable) ─

/** action.config.billable — the money side. Mirrors ActionBillable in action.py. */
export interface ActionBillable {
  // who is billed
  customer_id?: number | null;             // billing party (customer record)
  contact_id?: number | null;              // specific person on the account
  company?: string;                        // display name (denormalized)
  attention?: string;                      // attention line on invoice

  // rate
  is_billable?: boolean;                   // default true
  rate?: number | null;                    // per rate_unit
  rate_unit?: string;                      // "hour" | "day" | "flat" | "unit"
  currency?: string;                       // default "USD"

  // categorization
  skill_category?: string;                 // journeyman_plumber, electrician, etc.
  activity?: string;                       // installation, service, consultation, etc.

  // estimates vs actuals
  hours_estimated?: number | null;
  hours_actual?: number | null;
  total_estimated?: number | null;         // rate * hours_estimated (or flat amount)
  total_actual?: number | null;            // rate * hours_actual (or flat amount)
  variance_pct?: number | null;            // ((actual - estimated) / estimated) * 100

  // product set
  product_set?: number[];                  // Item IDs that this labor applies to

  // invoice linkage
  invoice_id?: number | null;
  invoice_line?: number | null;

  // Alice learning signal
  estimate_source?: string | null;         // "alice" | "user" | "historical"
  estimate_confidence?: number | null;     // 1-10

  [key: string]: unknown;                  // extra="allow" — industry extensions
}

// ── action.config — mirrors ActionConfig ────────────────────────────────

export interface ActionConfig {
  times?: ActionTimes | null;
  billable?: ActionBillable | null;
  [key: string]: unknown;                  // ConfigBase allows model-specific fields
}

// ── Action record ───────────────────────────────────────────────────────

export interface Action {
  id: number;
  uuid?: string;
  ida?: string;
  dt_created?: number;
  dt_modified?: number;
  version?: number;
  is_active?: boolean;
  is_deleted?: boolean;
  is_archived?: boolean;
  action_id?: string;
  action?: Record<string, string>;
  description?: Record<string, string>;
  languages?: string[];
  assigned_to?: Array<{ id: number; name: string; email?: string }>;
  contact_id?: number;
  project_name?: string;
  project_id?: number;
  project_metadata?: Record<string, any>;
  linkage?: number;
  sequence?: number;
  kanban_column?: string;
  priority?: number;
  difficulty?: number;
  status?: string;
  percent_complete?: number;
  burndown?: number;
  dt_start?: number;
  dt_expected?: number;
  dt_deadline?: number;
  dt_completed?: number;
  dt_updated?: number;
  duration?: number;
  created_by?: Array<any>;
  updated_by?: Array<any>;
  expected_by?: Array<any>;
  due_by?: Array<any>;
  completed_by?: Array<any>;
  start_by?: Array<any>;
  end_by?: Array<any>;
  config?: ActionConfig;
  impact?: Record<string, any>;
  retrospection?: Record<string, any>;
  metadata?: Record<string, any>;
  refs?: Record<string, any>;
  prefs?: Record<string, any>;
  comments?: Record<string, any>;
  actions?: Record<string, any>;
  health_rating?: number;
}
